packages/google_maps_flutter/google_maps_flutter_web/lib/src/marker_clustering_js_interop.dart

```diff
@@ -0,0 +1,116 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:google_maps/google_maps.dart' as gmaps;
+import 'package:js/js.dart';
```

Comment:
This needs to be done with `package:web` and `dart:js_interop` instead. See docs:

* https://dart.dev/interop/js-interop

Some examples (take a look at the whole files, there's examples on how to extend an existing package:web type, and how to create a custom object with an external factory):

* https://github.com/flutter/packages/blob/9a2490419268365911191f9d8388f986111bcc24/packages/video_player/video_player_web/example/integration_test/pkg_web_tweaks.dart#L41
* https://github.com/flutter/packages/blob/9a2490419268365911191f9d8388f986111bcc24/packages/google_maps_flutter/google_maps_flutter_web/lib/src/map_styler.dart#L17

Reply:
@ditman thanks for the review.
I once tried using dart:js_interop instead of package:js but got trouble with [gmaps.Marker references on bindings](https://github.com/a14n/dart-google-maps/blob/master/lib/src/generated/google_maps_core.js.g.dart#L1611)

[google_maps](https://pub.dev/packages/google_maps) package, that still uses js-package (not js_interop). 
See: https://github.com/a14n/dart-google-maps/blob/master/pubspec.yaml#L10
And for example https://github.com/a14n/dart-google-maps/blob/master/lib/src/generated/google_maps_core.js.g.dart#L1611 for marker bindings.

I am not sure is it possible to use js bindings from the google_maps package inside the js_interop extensions.
I can try again, but I don't think it is possible. So I think the  [google_maps](https://pub.dev/packages/google_maps) package need to be updated to use js_interop before this clustering support can land for the web platform.

I'll try again in separate branch and update the state.

Reply:
@jokerttu I think the current definition should allow extending the gmaps.Marker. If you prepare a branch but you encounter errors, do cc me so I can try and help!

The goal is to **not use `package:js` in this package anymore.** Note that the JS annotation is now also available from `dart:js_interop` ([doc](https://api.flutter.dev/flutter/dart-js_interop/JS-class.html)).

Reply:
@ditman here is the commit to separate branch:
https://github.com/CodemateLtd/packages/commit/3770ee2fcd3574b3320c16193e1bf835c0685b84

I was not yet able to solve the issue with the google_maps library:
example:
```
../../google_maps_flutter_web/lib/src/marker_clustering_js_interop.dart:22:28: Error: Type 'GMap?' is not a valid type in the signature of `dart:js_interop` external APIs or APIs converted via `toJS`. The only valid types are: JS types from `dart:js_interop`, @staticInterop types, void, bool, num, double, int, String, extension types that erases to one of these types, or a type parameter that is bound to a static interop type.
 - 'GMap' is from 'package:google_maps/src/generated/google_maps_core.js.g.dart' ('../../../../../../../.pub-cache/hosted/pub.dev/google_maps-7.1.0/lib/src/generated/google_maps_core.js.g.dart').
Use one of the valid types instead.
  external gmaps.GMap? get map;
```

Reply:
(I'm looking at this right now, will report back!)

Reply:
Fixed, link to PR below!

Reply:
@ditman thanks, looks great. Merged and this PR is updated.
After reading your PR it is really clear how the type conversion should be done with external JS types;
Even when this is actually explained in https://dart.dev/interop/js-interop/usage and https://dart.dev/interop/js-interop/js-types it would be good to have better example, at the type conversion chapter, of the case where external javascript types are used as a function parameters, as in this case.
Thanks again!